oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/config/group/EndpointNameGrouping.java

```diff
@@ -18,32 +18,91 @@
+import org.apache.skywalking.oap.server.core.config.group.uri.quickmatch.QuickUriGroupingRule;
+import org.apache.skywalking.oap.server.core.query.MetadataQueryService;
+import org.apache.skywalking.oap.server.library.util.CollectionUtils;
+import org.apache.skywalking.oap.server.library.util.RunnableWithExceptionProtection;
+import org.apache.skywalking.oap.server.library.util.StringFormatGroup;
 
 @Slf4j
 public class EndpointNameGrouping {
     @Setter
-    private volatile EndpointGroupingRule endpointGroupingRule;
+    private volatile QuickUriGroupingRule endpointGroupingRule;
     @Setter
     private volatile EndpointGroupingRule4Openapi endpointGroupingRule4Openapi;
+    @Setter
+    private volatile QuickUriGroupingRule quickUriGroupingRule;
+    /**
+     * Cache the HTTP URIs which are not formatted by the rules per service.
+     * Level one map key is service name, the value is a map of HTTP URI and its count.
+     * Multiple matches will be counted, because it is the pattern that the endpoint is already formatted,
+     * or doesn't need to be formatted.
+     * The repeatable URI is a pattern already.
+     */
+    private ConcurrentHashMap<String, ConcurrentHashMap<String, AtomicInteger>> cachedHttpUris = new ConcurrentHashMap<>();
+    private final AtomicInteger aiPipelineExecutionCounter = new AtomicInteger(0);
+    /**
+     * The max number of HTTP URIs per service for further URI pattern recognition.
+     */
+    private int maxHttpUrisNumberPerService = 3000;
 
-    public String format(String serviceName, String endpointName) {
-        String formattedName = endpointName;
+    /**
+     * Format the endpoint name according to the API patterns.
+     *
+     * @param serviceName  service name
+     * @param endpointName endpoint name to be formatted.
+     * @return Tuple2 where the first element is the formatted name and the second element is a boolean
+     * represented the endpoint name is formatted or not.
+     */
+    public Tuple2<String, Boolean> format(String serviceName, String endpointName) {
+        Tuple2<String, Boolean> formattedName = new Tuple2<>(endpointName, Boolean.FALSE);
         if (endpointGroupingRule4Openapi != null) {
-            formattedName = formatByOpenapi(serviceName, formattedName);
+            formattedName = formatByOpenapi(serviceName, endpointName);
+        }
+
+        if (!formattedName._2() && endpointGroupingRule != null) {
+            formattedName = formatByCustom(serviceName, endpointName);
+        }
+
+        if (!formattedName._2() && quickUriGroupingRule != null) {
+            formattedName = formatByQuickUriPattern(serviceName, endpointName);
         }
 
-        if (endpointGroupingRule != null) {
-            formattedName = formatByCustom(serviceName, formattedName);
+        if (!formattedName._2()) {
+            // Only URI starts with '/' will be cached and formatted later.
+            if (endpointName.indexOf("/") > -1) {
```

Comment:
`endpointName.indexOf("/") > -1` doesn't match the comment `Only URI starts with '/' will be cached and formatted later.`

```suggestion
            // Only URI starts with '/' will be cached and formatted later.
            if (endpointName.startsWith("/")) {
```

Reply:
Let me fix codes. start with is wrong. Get and Post are prefix, not /. That was a legacy comment.

Reply:
Fixed